metapool/metapool.py

```diff
@@ -480,6 +480,147 @@ def read_pico_csv(f, sep='\t', plate_reader='SpectraMax_i3x',
+
+
+def select_sample_dilutions(
+        a_plate_df, dilution_suffixes_in_order, a_mask_func,
+        a_col_name_func=None):
+    """
+    Selects the correct source of each sample based on input dilutions
+
+    Parameters:
+    a_plate_df : pandas DataFrame
+        A DataFrame of plate info, containing at least PM_WELL_KEY,
+        PM_PROJECT_PLATE_KEY, and PM_COMPRESSED_PLATE_NAME_KEY columns
+    dilution_suffixes_in_order : list of str
+        List of dilution suffix(es) in order from most to least preferred for
+        use. NOTE THAT the suffix for the base dilution plate must be the
+        LAST ONE.  The base dilution plate doesn't have to be *undiluted*,
+        but it must be the one that will be used when no other dilution meets
+        the user's criteria.
+    a_mask_func : function
+        A function to generate a mask that is true for samples that meet the
+        user's (arbitrary) criteria, taking as parameters a DataFrame and a
+        particular dilution's (full) concentration column name.
+    a_col_name_func : function
+        A function to generate the (existing) column name of a concentration
+        column in the plate dataframe based on its dilution suffix. If None,
+        defaults to SAMPLE_DNA_CONC_KEY_<suffix>, which matches what is
+        produced by _read_and_label_pico_csv.
+
+    Returns:
+    a_plate_df : pandas DataFrame
+        An expansion of the input plate_df; for each sample, the concentration
+        column is set to the sample's concentration in the most preferred
+        concentration that meets the user's criteria. Likewise, the plate name
+        and project plate columns are set to those for the selected dilution,
+        and the 'Diluted' column is set to true if the selected dilution is
+        NOT the base dilution.
+    """
+
+    if PM_DILUTED_KEY in a_plate_df.columns:
+        # function has already been applied to data
+        warnings.warn('Dilution operation was already performed')
+        return a_plate_df
+
+    def _default_col_name_func(x):
+        return f"{SAMPLE_DNA_CONC_KEY}_{x}"
+
+    a_col_name_func = _default_col_name_func if a_col_name_func is None \
+        else a_col_name_func
+
+    # return a copy of the input data. Do not overwrite input data by default.
+    df = a_plate_df.copy()
+
+    # first time through the loop is the base dilution case
+    is_base_dilution = True
+
+    # loop over the dictionary keys/items in reverse order
+    reverse_order_keys = list(reversed(dilution_suffixes_in_order))
+    for curr_dilution_suffix in reverse_order_keys:
+        curr_conc_key = a_col_name_func(curr_dilution_suffix)
+        curr_is_base_dilution = is_base_dilution
+        if is_base_dilution:
+            # make a mask including ALL rows; this is the fallback case
+            dilution_mask = pd.Series(True, index=df.index)
+            # going forward, all future suffixes will be for further dilutions
+            is_base_dilution = False
```

Comment:
Based on line 607, should we have a break here? I'm not sure why we need a "flag" like is_base_dilution? Maybe the confusion comes from line 607

Reply:
We don't want a break here because we do want to go on and finish the loop for the base dilution case, but we also want to know that the *next* time through the loop will NOT be the base dilution (and thus will get some different logic).  I have added expanded and more specific comments for this line to hopefully make it clear what this is doing.